### PR TITLE
Improve hexo type definition

### PR DESCRIPTION
**Pull request by KentarouTakeda**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
    * https://github.com/hexojs/hexo/blob/92b979f4a3fa8714aebd3d11c3295d466b870905/lib/hexo/render.js#L75
    * https://github.com/hexojs/hexo/blob/92b979f4a3fa8714aebd3d11c3295d466b870905/lib/hexo/render.js#L109
    * https://github.com/hexojs/hexo/blob/92b979f4a3fa8714aebd3d11c3295d466b870905/lib/hexo/index.js#L420
    * https://github.com/hexojs/hexo/blob/92b979f4a3fa8714aebd3d11c3295d466b870905/lib/hexo/index.js#L427
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/hexo/index.d.ts [hunk 1]
```diff
@@ -667,7 +667,12 @@ declare namespace Hexo {
         }
 
         interface Generator {
-            register(name: string, fn: (locals: Site) => Generator.Return | Generator.Return[]): void;
+            register(
+                name: string,
+                fn: (
+                    locals: Site,
+                ) => Generator.Return | Generator.Return[] | Promise<Generator.Return> | Promise<Generator.Return[]>,
```
Comment (domdomegg) [inline types/hexo/index.d.ts:674]: I'm not sure this does work. Under the hood, Hexo calls bluebird's `Promise.method` on the function which I think results in the error:

```
Uncaught TypeError: expecting a function but got [object Promise]
```

https://github.com/hexojs/hexo/blob/92b979f4a3fa8714aebd3d11c3295d466b870905/lib/extend/generator.js#L29

**Review by domdomegg — CHANGES_REQUESTED** [review]

File: types/hexo/index.d.ts [hunk 2]
```diff
@@ -704,16 +709,20 @@ declare namespace Hexo {
...
-                fn: (data: RendererData, options: any) => string,
+                fn: (this: Hexo, data: RendererData, options: any) => string,
                 sync: true,
             ): void;
             register(
                 srcExt: string,
                 outExt: string,
-                fn: (data: RendererData, options: any) => Promise<string>,
+                fn: (this: Hexo, data: RendererData, options: any) => Promise<string>,
                 sync: false,
```
[the remark this answers is not in the document]
Comment (KentarouTakeda) [reply types/hexo/index.d.ts]: Fixed. thank you!

File: types/hexo/index.d.ts (hunk 1 above)
Comment (KentarouTakeda) [reply types/hexo/index.d.ts]: Fixed. thank you!

**Review by domdomegg — APPROVED** [review]

**Comment by KentarouTakeda** [comment]
Ready to merge

**Pull request merged**